Document `VECTOR` type and new vector functions (#1299)
neo4j/docs-cypher: modules/ROOT/pages/values-and-types/vector.adoc

= Vectors
:description: Information about Cypher's `VECTOR` type.
:page-role: new-2025.xx

Cypher supports the construction of a `VECTOR` value that can be stored as xref:indexes/semantic-indexes/vector-indexes.adoc#embeddings[embedding] properties on nodes and relationships and utilized for efficient semantic retrieval using Neo4j's xref:indexes/semantic-indexes/vector-indexes.adoc[vector indexes] and xref:genai-integrations.adoc[GenAI plugin].
`VECTOR` values can also be measured and compared (in terms of similarity, distance, and norm) using Cypher's xref:functions/vector.adoc[Vector functions].

[[vector-type]]
== The vector type

> The `VECTOR` type is a fixed-length, ordered collection of of numeric coordinate values (`INTEGER` or `FLOAT`) stored as a single unit.

Reviewer: ```suggestion
The `VECTOR` type is a fixed-length, ordered collection of numeric coordinate values (`INTEGER` or `FLOAT`) stored as a single unit.
```

The following query constructs a three-dimensional vector with `INTEGER` values.
It returns the `VECTOR` itself, and the xref:functions/scalar.adoc#functions-valuetype[`valueType()`] function is used to return information about its type.

.Construct a `VECTOR` value
[source, cypher]
----
WITH vector([1, 2, 3], 3, INTEGER) AS vector
RETURN vector, valueType(vector) AS vectorType
----

.Result
[role="queryresult",options="header,footer",cols="2*<m"]
|===
| vector | vectorType

| [1, 2, 3] | "VECTOR<INTEGER NOT NULL>(3) NOT NULL"

2+d|Rows: 1
|===

The following example uses a parameter containing a `LIST<INTEGER>` as input to `vectorValue`.

.Parameters
[source, parameters]
----
{
  "integerList": [1, 2, 3, 4, 5]
}
----

.Construct a `VECTOR` value with using a parameter for `vectorValue`
[source, cypher]
----
WITH vector($integerList, 5, INTEGER8) as vector
RETURN vector, valueType(vector) AS vectorType
----

.Result
[role="queryresult",options="header,footer",cols="2*<m"]
|===
| vector | vectorType

| [1, 2, 3, 4, 5] | "VECTOR<INTEGER8 NOT NULL>(5) NOT NULL"

2+d|Rows: 1
|===

[role=label--enterprise-edition]
[[store-vector-properties]]

> == Store vector values as properties

Reviewer: It is worth repeating the following in this section:

```
| `VECTOR` values can be stored as properties if the instance uses Neo4j's link:{neo4j-docs-base-uri}/operations-manual/current/database-internals/store-formats/#store-format-overview[block format].
| `VECTOR` values cannot be stored as properties using Community Edition.
```

Author: @sherfert -- the header has the "Enterprise Edition" label (meaning ee only) and below (line 121-122) is a note about store format.

= Vectors
:description: Information about Cypher's `VECTOR` type.
:page-role: new-2025.xx

> Cypher's supports the construction of a `VECTOR` type that can be stored as xref:indexes/semantic-indexes/vector-indexes.adoc#embeddings[embedding] properties on nodes and relationships and utilized for efficient semantic retrieval using Neo4j's xref:indexes/semantic-indexes/vector-indexes.adoc[vector indexes] and xref:genai-integrations.adoc[GenAI plugin].

Reviewer 2: ```suggestion
Cypher supports the construction of a `VECTOR` value that can be stored as xref:indexes/semantic-indexes/vector-indexes.adoc#embeddings[embedding] properties on nodes and relationships and utilized for efficient semantic retrieval using Neo4j's xref:indexes/semantic-indexes/vector-indexes.adoc[vector indexes] and xref:genai-integrations.adoc[GenAI plugin].
```

`VECTOR` values can also be measured and compared (in terms of similarity, distance, and norm) using Cypher's xref:functions/vector.adoc[Vector functions].

[[vector-type]]
== The VECTOR type

The `VECTOR` type is a fixed-length list of numeric coordinate values (`INTEGER` or `FLOAT`) stored as a single unit.
It is defined by its dimension, which sets how many numeric values it holds, and its coordinate type, which specifies the specific numeric data type for those numbers.

> Each numeric value in the list represents a coordinate along one of the vector’s defined dimensions.

Reviewer 2: I think we need to be careful with the word list in this paragraph, a vector is not a list in the programming sense

The coordinate type controls the type of each coordinate in the `VECTOR`, influencing precision and storage size.
A `VECTOR` value must have a dimension and a coordinate type.

.Example `VECTOR` type
[source]
----
VECTOR([1.05, 0.123, 5], 3, FLOAT32 NOT NULL)
----

In this example `VECTOR` type value, `[1.05, 0.123, 5]` is the coordinates of the `VECTOR`, `3` defines its dimensionality, and `FLOAT32` specifies the data type for each coordinate value.

.`VECTOR` and `LIST` values
[NOTE]
The dimensions and the formatting of a `VECTOR` is similar to a xref:values-and-types/lists.adoc[`LIST`] value.
However, unlike `LIST` values which contain elements that can be accessed individually with xref:expressions/list-expressions.adoc[list expressions], all operations on a `VECTOR` must operate on the entire `VECTOR`; it is not possible to access or slice individual dimensions of a `VECTOR` value.

> Additionally, a `LIST` cannot contain `VECTOR` values.

Reviewer 2: ```suggestion
```

A list can contain a vector, just not be stored

The dimension of a `VECTOR` value must be larger than `0` and less than or equal to `4096`. 

The following coordinate types are supported:

.Vector coordinate types
[options="header",cols="2*<m"]
|===
| Default name | Alias

| `FLOAT` | `FLOAT64`
| `FLOAT32` | 
| `INTEGER` | `INT`, `INT64`, `SIGNED INTEGER`, `INTEGER64`
| `INTEGER8` | `INT8`
| `INTEGER16`| `INT16`
| `INTEGER32` | `INT8`

|===

> `VECTOR` is a considered a supertype of which all combinations of `VECTOR<TYPE>(DIMENSION)` are a part.

Reviewer 2: ```suggestion
`VECTOR` is considered a supertype of which all combinations of `VECTOR<TYPE>(DIMENSION)` are a part.
```

This logic continues for `VECTOR` types that define only a coordinate type or a dimension.
A  `VECTOR` with only a defined dimension is a supertype of all `VECTOR` values of that dimension regardless of the coordinate type.
For example, `VECTOR(4)` is a supertype of `VECTOR<FLOAT>(4)` and `VECTOR<INT8>(4)`.
Similarly, a `VECTOR` with only a defined coordinate type is a supertype of all `VECTOR` values with that coordinate type regardless of the dimension.
For example, `VECTOR<INT>` is a supertype of `VECTOR<INT>(3)` and `VECTOR<INT>(1024)`.
All of these supertypes can be used in xref:expressions/predicates/type-predicate-expressions.adoc#type-predicate-vector[type predicate expressions] to verify `VECTOR` type values.

[[construct-vector-values]]
== Construct VECTOR values

To construct a `VECTOR` value, use the xref:functions/vector.adoc#functions-vector[`vector()`] function.

The following query constructs a three-dimensional vector with `INTEGER` values.

> It returns the `VECTOR` itself, and the xref:functions/scalar.adoc#functions-valuetype[`valueType()`] function is used to information about its type.

Reviewer 2: ```suggestion
It returns the `VECTOR` itself, and the xref:functions/scalar.adoc#functions-valuetype[`valueType()`] function is used to return information about its type.
```

.Construct a `VECTOR` value
[source, cypher]
----
WITH vector([1, 2, 3], 3, INTEGER) AS vector
RETURN vector, valueType(vector) AS vectorType
----

.Result
[role="queryresult",options="header,footer",cols="2*<m"]
|===
| vector | vectorType

> | [1, 2, 3] | "VECTOR<INTEGER64 NOT NULL>(3) NOT NULL"

Reviewer 2: ```suggestion
| [1, 2, 3] | "VECTOR<INTEGER NOT NULL>(3) NOT NULL"
```

2+d|Rows: 1
|===

The following example uses a parameter containing a `LIST<INTEGER>` as input to `vectorValue`.

.Parameters
[source, parameters]
----
{
  "integerList": [1, 2, 3, 4, 5]
}
----

.Construct a `VECTOR` value with using a parameter for `vectorValue`
[source, cypher]
----
WITH vector($integerList, 5, INTEGER8) as vector
RETURN vector, valueType(vector) AS vectorType
----

.Result
[role="queryresult",options="header,footer",cols="2*<m"]
|===
| vector | vectorType

| [1, 2, 3, 4, 5] | "VECTOR<INTEGER8 NOT NULL>(5) NOT NULL"

2+d|Rows: 1
|===

[role=label--enterprise-edition]
[[store-vector-properties]]
== Store VECTOR values as properties

> To store a `VECTOR` value as a node or relationship property, use the `vector()` function together with a write clause.

Reviewer 2: Maybe mention here that only vector values can be stored and not lists of them

.Create a node and assign a `VECTOR` property
[source, cypher]
----

> CREATE (n:Label {vectorProp = vector([1,2,3], 3, INTEGER)})

Reviewer 2: ```suggestion
CREATE (n:Label {vectorProp: vector([1,2,3], 3, INTEGER)})
```

RETURN n.vectorProp AS vectorProp
----

.Result
[role="queryresult",options="header,footer",cols="1*<m"]
|===
| vectorProp

| 

1+d|Rows: 1
|===

.Parameters
[source, parameters]
----
{
  "floatList": [1.3, 2.4, 3.1, 4.4, 5.9]
}
----

.Set a `VECTOR` property using a parameter
[source, cypher]
----
CREATE (n:Label)
SET n.vectorProp = vector($floatList, 5, FLOAT)
RETURN n.vectorProp AS vectorProp
----

.Result
[role="queryresult",options="header,footer",cols="1*<m"]
|===
| vectorProp

| 

1+d|Rows: 1
|===

[[genai-plugin-vector-indexes]]
== VECTOR values, the GenAI plugin, and vector indexes

> Neo4j previously stored vector embeddings as `LIST<FLOAT64>` property values.

Reviewer 2: Is this true? I thought that the procedure stored them more efficiently

[NOTE]
The dimensions and the formatting of a `VECTOR` is similar to a xref:values-and-types/lists.adoc[`LIST`] value.
However, unlike `LIST` values which contain elements that can be accessed individually with xref:expressions/list-expressions.adoc[list expressions], all operations on a `VECTOR` must operate on the entire `VECTOR`; it is not possible to access or slice individual dimensions of a `VECTOR` value.

> Additionally, a `LIST` cannot store `VECTOR` values.

Reviewer 2: Neo4j cannot store Lists containing vector values, slight difference
